src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -137,33 +138,41 @@ async Task AddDocumentsAndProject(Project? project, ImmutableArray<string> suppo
                 {
                     var sourceGeneratedDocuments = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
                     documents = documents.AddRange(sourceGeneratedDocuments);
+                    todoCommentDocuments = todoCommentDocuments.AddRange(sourceGeneratedDocuments);
                 }
 
-                foreach (var document in documents)
-                {
-                    // Only consider closed documents here (and only open ones in the DocumentPullDiagnosticHandler).
-                    // Each handler treats those as separate worlds that they are responsible for.
-                    if (context.IsTracking(document.GetURI()))
-                    {
-                        context.TraceInformation($"Skipping tracked document: {document.GetURI()}");
-                        continue;
-                    }
-
-                    // Do not attempt to get workspace diagnostics for Razor files, Razor will directly ask us for document diagnostics
-                    // for any razor file they are interested in.
-                    if (document.IsRazorDocument())
-                    {
-                        continue;
-                    }
-
-                    result.Add(new WorkspaceDocumentDiagnosticSource(document));
-                }
+                // Scan for todo-comments first.  It's syntactic and much faster than semantic analysis.
+                AddSourceIfTracked(todoCommentDocuments, static d => d is Document doc ? new TodoCommentDiagnosticSource(doc) : null);
```

Comment:
I just thought of this, but it seems like a bit of a waste to make a whole call to OOP for all of these documents if the task list isn't even showing.  I wonder if the client should have some notion of kinds it is requesting in the request.  

This'll also end up being a good amount more work for workspace pull especially when FSA is off (as currently it'll only get diagnostics for open /active docs).  That's going to be a ton of OOP requests for a solution like Roslyn (even if their fast).  I'm leaning towards having todo comment analysis just support the FSA option as well

Reply:
Indeed. I'll bring up with Hannah and Kayle!

Reply:
> I'm leaning towards having todo comment analysis just support the FSA option as well

So this would be a regression on how TODO currently works today (it runs for the entire solution).  I don't think it would be ok to regress that.  we do have users using this as a critical part of how they do work (we know because we broke it and we got a lot of angry users over it :)).